Modified URDF parser to support URDFs that include the world as a link.

liangfok opened the pull request:
Closes: https://github.com/RobotLocomotion/drake/issues/2168
Contributes to: https://github.com/RobotLocomotion/drake/pull/2163

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/2171) &emsp; Multiple assignees:&emsp;<img alt="@RussTedrake" height="20" width="20" align="absmiddle" src="https://avatars.githubusercontent.com/u/6442292?s=40&v=3">&nbsp;<a href="/robotlocomotion/drake/pulls/assigned/RussTedrake">RussTedrake</a>,&emsp;<img alt="@ggould-tri" height="20" width="20" align="absmiddle" src="https://avatars.githubusercontent.com/u/17596413?s=40&v=3">&nbsp;<a href="/robotlocomotion/drake/pulls/assigned/ggould-tri">ggould-tri</a>,&emsp;<img alt="@liangfok" height="20" width="20" align="absmiddle" src="https://avatars.githubusercontent.com/u/1388098?s=40&v=3">&nbsp;<a href="/robotlocomotion/drake/pulls/assigned/liangfok">liangfok</a>
<!-- Reviewable:end -->

RussTedrake commented:
looks good to me.  thanks for doing it quickly.  

do we need to implement the same functionality for SDF?  (in a second PR)?

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 684 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KGBUNUIn163skU02aZI:-KGBUNUJNpCAMv8tYMAf:-24738346) ([raw file](https://github.com/robotlocomotion/drake/blob/8ff3a4b4b4cc996b6a95f5632d677b2bea27b603/drake/systems/plants/RigidBodyTreeURDF.cpp#L684)):*
nit: specified

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-:-KGBUmijWb4MK_6MeBwD:-1031717093)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@russtedrake, this is ready for another round of review.

+@russtedrake 

---

Review status: 2 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/plants/test/load_model_test.cc, line 277 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KGCPBhfWUGT3L0VwKaF:-KGCawz0xkMOEx33H_th:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/8ff3a4b4b4cc996b6a95f5632d677b2bea27b603/drake/systems/plants/test/load_model_test.cc#L277)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-:-KGCaoVEewIjy8X92kUy:1594918546)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-:-KGCn3p_tRtlKg9buB-A:1417088721)*
<!-- Sent from Reviewable.io -->

liangfok commented:
As a workaround to https://github.com/RobotLocomotion/drake/issues/2175, I just updated this PR by moving the model files used by `load_model_test` into a separate `models/` subdirectory. I think all model files in this `test/` directory should eventually go there.

RussTedrake commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 17 of 17 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-:-KGFEXYFodGdAIr8aQEs:1417088721)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@jwnimmer-tri: I don't have bits to override CI failure to merge. Thus, assigning this to you.

jwnimmer-tri commented:
@liangfok Jenkins on 4281804 was a failure on linux_matlab that looks real.  I am not seeing why I should override-merge that.

liangfok commented:
@jwnimmer-tri: I'm unable to view the error, see below. How did you pull it up?

![screenshot from 2016-04-25 22 22 21](https://cloud.githubusercontent.com/assets/1388098/14804782/3fb9b20a-0b34-11e6-9be9-1998005ebd17.png)

jwnimmer-tri commented:
Github shows the list of pushes to the branch.  The ones that end up starting a jenkins build have a jenkins marker next to them, as in the two red Xs here:
![screenshot from 2016-04-25 22 27 58](https://cloud.githubusercontent.com/assets/17596505/14804882/fe2193b6-0b34-11e6-8ab8-666218bf57ec.png)
If you click on the red X, you get a pop-up with a link to jenkins.

The only time that doesn't work is if you force-push a branch, the links to old jenkins runs get nuked too.  Then you have to dig around in the build list searching for your PR number.

jwnimmer-tri commented:
And in this case, the most recent jenkins output is here:
https://drake-jenkins.csail.mit.edu/job/experimental/1034/

liangfok commented:
Ah, I was selecting the wrong CI server (drake005 vs. Jenkins). The problem is caused by https://github.com/RobotLocomotion/drake/issues/2175. Looks like the Matlab unit test is able to find the URDFs with a world link even though I tried to hide them within a `test/models/` subdirectory.

I'll work out a solution and ping you again.

liangfok commented:
This PR is currently blocked until https://github.com/RobotLocomotion/drake/pull/2221 is merged.

jwnimmer-tri commented:
-@jwnimmer-tri as reviewer.  Feel free to add me back once this is unblocked.  (Reviewable can't filter on "do not merge".)

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> This PR is currently blocked until https://github.com/RobotLocomotion/drake/pull/2221 is merged.</details>


---

Review status: 9 of 10 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-:-KHRbU3mPn2qqS5HgYl0:-1821422272)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I am going to mark this PR "status: do not review" until after PR https://github.com/RobotLocomotion/drake/pull/2356 is merge.

liangfok commented:
https://github.com/RobotLocomotion/drake/pull/2356 was merged. Thus, I believe this PR can be retested. I shall update this branch with the latest from master, which should trigger another round of CI tests.

liangfok commented:
The CI test infrastructure detected a problem on Windows 32:
https://drake-cdash.csail.mit.edu/testDetails.php?test=673200&build=21149

I will look into it soon.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> https://github.com/RobotLocomotion/drake/pull/2356 was merged. Thus, I believe this PR can be retested. I shall update this branch with the latest from master, which should trigger another round of CI tests.</details>


---

Review status: 8 of 11 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-:-KHzUkmFWF7Kxpvbxxg_:-761744479)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I successfully replicated this problem on my local Windows machine. Onward towards the the debugging process...

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> The CI test infrastructure detected a problem on Windows 32:
> https://drake-cdash.csail.mit.edu/testDetails.php?test=673200&build=21149
> 
> I will look into it soon.
</details>


---

Review status: 8 of 11 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-:-KI2tOYiMpsVTYTEsowc:692078207)*
<!-- Sent from Reviewable.io -->

liangfok commented:
The error is transient. As a random sample, I ran the following unit test on Windows 32-bit 10 times and observed the problem 7 times.

```
$ ctest -VV -R load_model_test -C RelWithDebInfo
```

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I successfully replicated this problem on my local Windows machine. Onward towards the the debugging process...
</details>


---

Review status: 8 of 11 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-:-KI6FiGAJgofogpO9glU:-1702647366)*
<!-- Sent from Reviewable.io -->

liangfok commented:
For the record, here is an example of the transient error on 32-bit Windows machines:

```
25: unknown file: error: SEH exception with code 0xc0000005 thrown in the test body.
25: [  FAILED  ] MODEL_TO_WORLD_TRANSFORM_TESTS/ModelToWorldTransformTest.TestModelToWorldTransform/0, where GetParam() = 96-byte object <08-22 63-00 5C-00 57-00 69-00 6E-00 64-00 6F-00 64-00 00-00 6F-00 00-00 6C-69 6E-6B 31-00 65-00 6D-00 33-00 32-00 5C-00 05-00 00-00 0F-00 00-00 00-00 00-00 00-00 F0-3F 00-00 00-00 00-00 00-40 00-00 00-00 00-00 08-40 9A-99 99-99 99-99 B9-3F 00-00 00-00 00-00 E0-3F CD-CC CC-CC CC-CC FC-3F> (22 ms)
```

I first narrowed down the problem to the following lines in the `load_model_test` unit test:

```
  EXPECT_TRUE(
      link1_body->getJoint().getTransformToParentBody().matrix().isApprox(
          T_model_to_world.matrix(), 1e-10))
      << "Incorrect transform from the link1's frame to Drake's world frame."
      << "Got:\n"
      << link1_body->getJoint().getTransformToParentBody().matrix() << "\n"
      << "Expected:\n"
      << T_model_to_world.matrix();
```

I then further narrowed down the problem to the following code: `Isometry3d::matrix().isApprox(Isometry3d::matrix())`.

If I save the return values of `Isometry3d::matrix()` into local variables and then use the the local variables in the call to `isApprox()`, the error goes away.

I'm currently removing scaffolding code that I used to determine the above; hopefully the problem will not return as I remove this scaffolding.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> The error is transient. As a random sample, I ran the following unit test on Windows 32-bit 10 times and observed the problem 7 times.
> 
> ```
> $ ctest -VV -R load_model_test -C RelWithDebInfo
> ```
</details>


---

Review status: 8 of 11 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-:-KI9IhnTs8W4eV6yDIMO:601362889)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I believe I resolved the problem on 32-bit Windows problem. Hopefully the latest CI built will pass.

Unfortunately, I was not able to fully understand why the problem occurred. I simply found a workaround.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> For the record, here is an example of the transient error on 32-bit Windows machines:
> 
> ```
> 25: unknown file: error: SEH exception with code 0xc0000005 thrown in the test body.
> 25: [  FAILED  ] MODEL_TO_WORLD_TRANSFORM_TESTS/ModelToWorldTransformTest.TestModelToWorldTransform/0, where GetParam() = 96-byte object <08-22 63-00 5C-00 57-00 69-00 6E-00 64-00 6F-00 64-00 00-00 6F-00 00-00 6C-69 6E-6B 31-00 65-00 6D-00 33-00 32-00 5C-00 05-00 00-00 0F-00 00-00 00-00 00-00 00-00 F0-3F 00-00 00-00 00-00 00-40 00-00 00-00 00-00 08-40 9A-99 99-99 99-99 B9-3F 00-00 00-00 00-00 E0-3F CD-CC CC-CC CC-CC FC-3F> (22 ms)
> ```
> 
> I first narrowed down the problem to the following lines in the `load_model_test` unit test:
> 
> ```
>   EXPECT_TRUE(
>       link1_body->getJoint().getTransformToParentBody().matrix().isApprox(
>           T_model_to_world.matrix(), 1e-10))
>       << "Incorrect transform from the link1's frame to Drake's world frame."
>       << "Got:\n"
>       << link1_body->getJoint().getTransformToParentBody().matrix() << "\n"
>       << "Expected:\n"
>       << T_model_to_world.matrix();
> ```
> 
> I then further narrowed down the problem to the following code: `Isometry3d::matrix().isApprox(Isometry3d::matrix())`.
> 
> If I save the return values of `Isometry3d::matrix()` into local variables and then use the the local variables in the call to `isApprox()`, the error goes away.
> 
> I'm currently removing scaffolding code that I used to determine the above; hopefully the problem will not return as I remove this scaffolding.
</details>


---

Review status: 8 of 13 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-:-KI9iU90hViB0lUn0BQW:527736377)*
<!-- Sent from Reviewable.io -->

liangfok commented:
[Jenkins Build #1585](https://drake-jenkins.csail.mit.edu/job/experimental/1585/) seems to have suffered a strange Matlab flake. @jwnimmer-tri can you take a look?

jwnimmer-tri commented:
Sorry, I am stumped.

liangfok commented:
The previous build error appears to truly be a flakey since it passed the second time around.  

https://drake-jenkins.csail.mit.edu/job/experimental/1591/

This, thus PR is ready for final review/merge.

jwnimmer-tri commented:
@liangfok To be clear, we have Russ as feature LGTM, and you'd like me to platform review?

liangfok commented:
@jwnimmer-tri: Yes please. Thanks!

jwnimmer-tri commented:
+@ggould-tri for platform review (I haven't even started, so it turns into a Friday review).

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> @jwnimmer-tri: Yes please. Thanks!
</details>


---

Review status: 8 of 13 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-:-KIB12BHelH_WYdYqrQX:371148441)*
<!-- Sent from Reviewable.io -->

liangfok commented:
OK thanks!

<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> +@ggould-tri for platform review (I haven't even started, so it turns into a Friday review).
</details>


---

Review status: 8 of 13 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-:-KIB1l044grQnqzGBn6v:693288697)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
added my 2 cents

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> OK thanks!
</details>


---

Reviewed 1 of 17 files at r3, 1 of 3 files at r6, 1 of 1 files at r7, 2 of 3 files at r9, 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/plants/test/load_model_test.cc, line 331 \[r10\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KICaZsbVFewwwhO5PTi:-KICaZsbVFewwwhO5PTj:-876059084) ([raw file](https://github.com/robotlocomotion/drake/blob/424db8ca06c5989a715a52dfe48ef07526c4029f/drake/systems/plants/test/load_model_test.cc#L331)):*
> ```cpp
>   // see:
>   //
>   // https://github.com/robotlocomotion/drake/pull/2171#issuecomment-219770037
> ```

I can't see why, exactly, but I suspect that a call to eval() might have helped.  i recommend reading through https://eigen.tuxfamily.org/dox-devel/group__TopicAliasing.html if you haven't.

---

*[drake/util/allURDFs.m, line 24 \[r10\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KICb9F2ladWQkaNx0-o:-KICb9F3fhFJIF6UIHZC:-240023345) ([raw file](https://github.com/robotlocomotion/drake/blob/424db8ca06c5989a715a52dfe48ef07526c4029f/drake/util/allURDFs.m#L24)):*
> ```objc
> % Updates the above command to blacklist any URDF whose full path contains the
> % string "fixed_to_world".
> command = [command, '-not -path "', rootdir, '*fixed_to_world*" '];
> ```

does the existence of the world link break all matlab urdf parsing, or just the compareMex test?  if only in the comparison, i would blacklist it in compareMexTest instead, and explicitly use the existence of the world link as the defining feature, not the pathname.  (compareMexTest already skips past other urdfs based on features that don't match, too)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-:-KICbXO_HP3-tCNZ2W8M:-1412836850)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/plants/test/load_model_test.cc, line 331 \[r10\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KICaZsbVFewwwhO5PTi:-KICkfq34-5BJFTNRy9j:207253855) ([raw file](https://github.com/robotlocomotion/drake/blob/424db8ca06c5989a715a52dfe48ef07526c4029f/drake/systems/plants/test/load_model_test.cc#L331)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> I can't see why, exactly, but I suspect that a call to eval() might have helped.  i recommend reading through https://eigen.tuxfamily.org/dox-devel/group__TopicAliasing.html if you haven't.
</details>
I remember reading that Eigen aliasing article a few years ago; it's good to refresh my memory by reading it again. Unfortunately, I tried using it as follows and the transient error returned:

```
  EXPECT_TRUE(link1_body->getJoint().getTransformToParentBody().matrix().eval()
    .isApprox(T_model_to_world.matrix().eval(), 1e-10))
    << "ERROR: "
    << "Incorrect transform from the link1's frame to Drake's world frame.\n"
    << "Got:\n"
    << link1_body->getJoint().getTransformToParentBody().matrix().eval() << "\n"
    << "Expected:\n"
    << T_model_to_world.matrix().eval();
```
Here is another example of the error, which occurs in most but not every execution of the unit test:

```
$ ctest -VV -C RelWithDebInfo -R load_model_test
...
25: [----------] 2 tests from MODEL_TO_WORLD_TRANSFORM_TESTS/ModelToWorldTransformTest
25: [ RUN      ] MODEL_TO_WORLD_TRANSFORM_TESTS/ModelToWorldTransformTest.TestModelToWorldTransform/0
25: unknown file: error: SEH exception with code 0xc0000005 thrown in the test body.
25: [  FAILED  ] MODEL_TO_WORLD_TRANSFORM_TESTS/ModelToWorldTransformTest.TestModelToWorldTransform/0, where GetParam() = 96-byte object <70-62 D5-00 5C-00 57-00 69-00 6E-00 64-00 6F-00 64-00 00-00 6F-00 00-00 6C-69 6E-6B 31-00 65-00 6D-00 33-00 32-00 5C-00 05-00 00-00 0F-00 00-00 00-00 00-00 00-00 F0-3F 00-00 00-00 00-00 00-40 00-00 00-00 00-00 08-40 9A-99 99-99 99-99 B9-3F 00-00 00-00 00-00 E0-3F CD-CC CC-CC CC-CC FC-3F> (19 ms)
25: [ RUN      ] MODEL_TO_WORLD_TRANSFORM_TESTS/ModelToWorldTransformTest.TestModelToWorldTransform/1
25: unknown file: error: SEH exception with code 0xc0000005 thrown in the test body.
25: [  FAILED  ] MODEL_TO_WORLD_TRANSFORM_TESTS/ModelToWorldTransformTest.TestModelToWorldTransform/1, where GetParam() = 96-byte object <60-63 D5-00 5C-00 50-00 72-00 6F-00 67-00 72-00 67-00 00-00 6F-00 00-00 6C-69 6E-6B 31-00 73-00 5C-00 47-00 69-00 74-00 05-00 00-00 0F-00 00-00 00-00 00-00 00-00 08-40 00-00 00-00 00-00 00-40 00-00 00-00 00-00 F0-3F 9A-99 99-99 99-99 C9-3F CD-CC CC-CC CC-CC EC-3F 1F-85 EB-51 B8-1E F9-BF> (19 ms)
25: [----------] 2 tests from MODEL_TO_WORLD_TRANSFORM_TESTS/ModelToWorldTransformTest (38 ms total)
25:
25: [----------] Global test environment tear-down
25: [==========] 11 tests from 3 test cases ran. (245 ms total)
25: [  PASSED  ] 9 tests.
25: [  FAILED  ] 2 tests, listed below:
25: [  FAILED  ] MODEL_TO_WORLD_TRANSFORM_TESTS/ModelToWorldTransformTest.TestModelToWorldTransform/0, where GetParam() = 96-byte object <70-62 D5-00 5C-00 57-00 69-00 6E-00 64-00 6F-00 64-00 00-00 6F-00 00-00 6C-69 6E-6B 31-00 65-00 6D-00 33-00 32-00 5C-00 05-00 00-00 0F-00 00-00 00-00 00-00 00-00 F0-3F 00-00 00-00 00-00 00-40 00-00 00-00 00-00 08-40 9A-99 99-99 99-99 B9-3F 00-00 00-00 00-00 E0-3F CD-CC CC-CC CC-CC FC-3F>
25: [  FAILED  ] MODEL_TO_WORLD_TRANSFORM_TESTS/ModelToWorldTransformTest.TestModelToWorldTransform/1, where GetParam() = 96-byte object <60-63 D5-00 5C-00 50-00 72-00 6F-00 67-00 72-00 67-00 00-00 6F-00 00-00 6C-69 6E-6B 31-00 73-00 5C-00 47-00 69-00 74-00 05-00 00-00 0F-00 00-00 00-00 00-00 00-00 08-40 00-00 00-00 00-00 00-40 00-00 00-00 00-00 F0-3F 9A-99 99-99 99-99 C9-3F CD-CC CC-CC CC-CC EC-3F 1F-85 EB-51 B8-1E F9-BF>
25:
25:  2 FAILED TESTS
```
What a strange mystery!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Out of an abundance of caution, I modified the problematic `load_model_test` to allocate the `RigidBodySystem` on the heap to ensure it is 16-bit aligned.

<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> added my 2 cents
</details>


---

Review status: 12 of 13 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-:-KICtUgBUUU50V5Cq_t8:2072047636)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 12 of 13 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/util/allURDFs.m, line 24 \[r10\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KICb9F2ladWQkaNx0-o:-KICxtYzc00_sg-lYXxt:-328837667) ([raw file](https://github.com/robotlocomotion/drake/blob/424db8ca06c5989a715a52dfe48ef07526c4029f/drake/util/allURDFs.m#L24)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> does the existence of the world link break all matlab urdf parsing, or just the compareMex test?  if only in the comparison, i would blacklist it in compareMexTest instead, and explicitly use the existence of the world link as the defining feature, not the pathname.  (compareMexTest already skips past other urdfs based on features that don't match, too)
</details>
Good question. Testing here: https://github.com/RobotLocomotion/drake/pull/2412

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 12 of 13 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/util/allURDFs.m, line 24 \[r10\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KICb9F2ladWQkaNx0-o:-KID76wLsSTAFyFkZQkO:-894220070) ([raw file](https://github.com/robotlocomotion/drake/blob/424db8ca06c5989a715a52dfe48ef07526c4029f/drake/util/allURDFs.m#L24)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Good question. Testing here: https://github.com/RobotLocomotion/drake/pull/2412
</details>
The results indicate that the existence of the world link does in fact _only_ break the `systems/plants/test/compareParsers` unit test. 

https://drake-cdash.csail.mit.edu/viewTest.php?onlyfailed&buildid=23488

I will thus modify this PR to not black list world-link-containing URDFs in `allURDFs.m` and instead do that in compareMex Test.

@russtedrake: where is the compareMex test? I looked in [drake/systems/plants/test/compareParsersmex.cpp](https://github.com/RobotLocomotion/drake/blob/master/drake/systems/plants/test/compareParsersmex.cpp) and didn't see where it's currently blacklisting certain URDFs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: 12 of 13 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/util/allURDFs.m, line 24 \[r10\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KICb9F2ladWQkaNx0-o:-KIDDaQoHGv_KzjtfZ9T:-1797326157) ([raw file](https://github.com/robotlocomotion/drake/blob/424db8ca06c5989a715a52dfe48ef07526c4029f/drake/util/allURDFs.m#L24)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> The results indicate that the existence of the world link does in fact _only_ break the `systems/plants/test/compareParsers` unit test. 
> 
> https://drake-cdash.csail.mit.edu/viewTest.php?onlyfailed&buildid=23488
> 
> I will thus modify this PR to not black list world-link-containing URDFs in `allURDFs.m` and instead do that in compareMex Test.
> 
> @russtedrake: where is the compareMex test? I looked in [drake/systems/plants/test/compareParsersmex.cpp](https://github.com/RobotLocomotion/drake/blob/master/drake/systems/plants/test/compareParsersmex.cpp) and didn't see where it's currently blacklisting certain URDFs.
</details>
It's in the .m file that calls that mex file (same as the name of the failing test)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 11 of 14 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/util/allURDFs.m, line 24 \[r10\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KICb9F2ladWQkaNx0-o:-KID_nojpWDV735SWYrA:882828959) ([raw file](https://github.com/robotlocomotion/drake/blob/424db8ca06c5989a715a52dfe48ef07526c4029f/drake/util/allURDFs.m#L24)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> It's in the .m file that calls that mex file (same as the name of the failing test)
</details>
Thanks. I updated this PR to black list the world-containing URDFs in `compareParsers.m`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 1 of 3 files at r6, 1 of 3 files at r9, 1 of 1 files at r10, 1 of 1 files at r11, 2 of 3 files at r12.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTreeSDF.cpp, line 218 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KIDYnkszFwIjNsVeI_Q:-KIDYnktf0K0IOlXdXQY:1928604255) ([raw file](https://github.com/robotlocomotion/drake/blob/ceefe8a3a2422a4e665169fc22e2efca46e2aa1d/drake/systems/plants/RigidBodyTreeSDF.cpp#L218)):*
> ```cpp
>   if (!attr) {
>     throw runtime_error(
>       "RigidBodyTreeSDF.cpp: parseSDFLink: ERROR: "
> ```

FYI: Consider using `__FILE__` and `__func__` here so that the error will not become out-of-date if the file or function is renamed or the error code copy-pasted in the future.

e.g., `__FILE__ + ": " + __func__ + ": ERROR: Link tag..."`

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 398 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KID_rqTAZWHiZFpdGkD:-KID_rqTAZWHiZFpdGkE:-779390505) ([raw file](https://github.com/robotlocomotion/drake/blob/ceefe8a3a2422a4e665169fc22e2efca46e2aa1d/drake/systems/plants/RigidBodyTreeURDF.cpp#L398)):*
> ```cpp
>   if (!parent_node)
>     throw runtime_error("ERROR: joint " + name +
>                         " doesn't have a parent node!");
> ```

With this change, the exception texts in this file (indeed in this diff) are split between ending with "!", ".", and with no punctuation.

If you are going to change the punctuation of these exception texts, you should do so uniformly.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 589 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KIDb97qVPGfskaDO7UT:-KIDb97qVPGfskaDO7UU:-1998978021) ([raw file](https://github.com/robotlocomotion/drake/blob/ceefe8a3a2422a4e665169fc22e2efca46e2aa1d/drake/systems/plants/RigidBodyTreeURDF.cpp#L589)):*
> ```cpp
>  *
>  * An exception is thrown if no such joint is found, or if multiple
>  * world-connecting joints are found.
> ```

I was unable to find support or justification for this restriction in the URDF spec.  If this is a new restriction unique to Drake we should justify it.

(If it is a part of the URDF spec that I have failed to find, my apologies.)

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 594 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KIDcbNo8Hhz6__hTcoU:-KIDcbNo8Hhz6__hTcoV:147897355) ([raw file](https://github.com/robotlocomotion/drake/blob/ceefe8a3a2422a4e665169fc22e2efca46e2aa1d/drake/systems/plants/RigidBodyTreeURDF.cpp#L594)):*
> ```cpp
>  * @param[out] floating_base_type A reference to where the floating_base_type
>  * should be saved.
>  * @param[out] weld_to_frame The parameter to modify.
> ```

Please document the null behaviour here (ie, that it constructs a new RigidBodyFrame).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Reviewed 1 of 1 files at r11, 3 of 3 files at r12.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/plants/@RigidBodyManipulator/RigidBodyManipulator.m, line 344 \[r12\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KIF3OFlPZLAv1Njgmwn:-KIF3OFlPZLAv1Njgmwo:951978614) ([raw file](https://github.com/robotlocomotion/drake/blob/b645fd1d5ae0704779dda2640a0bab308ac7aa16/drake/systems/plants/@RigidBodyManipulator/RigidBodyManipulator.m#L344)):*
> ```objc
>     end
> 
>     function has_world_link = hasWorldLink(rbm)
> ```

You should really add test coverage for this method.  Could be as simple as loading one urdf with a workd link and one without and checking.

---

*[drake/systems/plants/@RigidBodyManipulator/RigidBodyManipulator.m, line 355 \[r12\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KIF2qS3p2IElolLD5nm:-KIF2qS4nJ6gZH8HfryF:321767803) ([raw file](https://github.com/robotlocomotion/drake/blob/b645fd1d5ae0704779dda2640a0bab308ac7aa16/drake/systems/plants/@RigidBodyManipulator/RigidBodyManipulator.m#L355)):*
> ```objc
>       % @ingroup Kinematic Tree
>       has_world_link = false;
>       for index = 1:length(rbm.body)
> ```

Note that this can be written to be more efficient without the for loop.  But it's fine for this PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@ggould-tri and @russtedrake: I've addressed your review comments. Thanks for the reviews. PTAL.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Out of an abundance of caution, I modified the problematic `load_model_test` to allocate the `RigidBodySystem` on the heap to ensure it is 16-bit aligned.
</details>


---

Review status: 10 of 16 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTreeSDF.cpp, line 218 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KIDYnkszFwIjNsVeI_Q:-KIDm-y-2hMUURes8-Wc:-909683839) ([raw file](https://github.com/robotlocomotion/drake/blob/ceefe8a3a2422a4e665169fc22e2efca46e2aa1d/drake/systems/plants/RigidBodyTreeSDF.cpp#L218)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> FYI: Consider using `__FILE__` and `__func__` here so that the error will not become out-of-date if the file or function is renamed or the error code copy-pasted in the future.
> 
> e.g., `__FILE__ + ": " + __func__ + ": ERROR: Link tag..."`
</details>
Fixed. I forgot about the availability of those macros. Thanks!

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 398 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KID_rqTAZWHiZFpdGkD:-KIFR9pjBlJmrE5gtW5l:-972462413) ([raw file](https://github.com/robotlocomotion/drake/blob/ceefe8a3a2422a4e665169fc22e2efca46e2aa1d/drake/systems/plants/RigidBodyTreeURDF.cpp#L398)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> With this change, the exception texts in this file (indeed in this diff) are split between ending with "!", ".", and with no punctuation.
> 
> If you are going to change the punctuation of these exception texts, you should do so uniformly.
</details>
Fixed. I updated _all_ error messages in this file to be uniform.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 589 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KIDb97qVPGfskaDO7UT:-KIFRoL4gtcytJGGZExE:-1257777640) ([raw file](https://github.com/robotlocomotion/drake/blob/ceefe8a3a2422a4e665169fc22e2efca46e2aa1d/drake/systems/plants/RigidBodyTreeURDF.cpp#L589)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> I was unable to find support or justification for this restriction in the URDF spec.  If this is a new restriction unique to Drake we should justify it.
> 
> (If it is a part of the URDF spec that I have failed to find, my apologies.)
</details>
I've updated the documentation.

It is an implicit specification of URDF. Each URDF file describes one robot [1] using a tree of links connected by joints. Thus, the only way for a URDF to contain two world-connecting joints is if the URDF describes more than a one robot. This is a violation of the one-robot-per-URDF rule.

[1] http://answers.gazebosim.org/question/62/sdf-vs-urdf-what-should-one-use/

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 594 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KIDcbNo8Hhz6__hTcoU:-KIFrY5ZoesvXGTxNzOj:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/ceefe8a3a2422a4e665169fc22e2efca46e2aa1d/drake/systems/plants/RigidBodyTreeURDF.cpp#L594)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> Please document the null behaviour here (ie, that it constructs a new RigidBodyFrame).
</details>
Done.

---

*[drake/systems/plants/@RigidBodyManipulator/RigidBodyManipulator.m, line 344 \[r12\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KIF3OFlPZLAv1Njgmwn:-KIGFWQcCvJ1XKyoFvFZ:1339939308) ([raw file](https://github.com/robotlocomotion/drake/blob/b645fd1d5ae0704779dda2640a0bab308ac7aa16/drake/systems/plants/@RigidBodyManipulator/RigidBodyManipulator.m#L344)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> You should really add test coverage for this method.  Could be as simple as loading one urdf with a workd link and one without and checking.
</details>
Done. This was actually a very good request because I discovered a major bug while writing the unit test: all `RigidBodyManipulator` objects have world links since one is added by the constructor. Thus, this method always returned `true`, meaning `compareParsers.m` skipped every URDF file!

I modified`RigidBodyManipulator` to throw an exception if a world link is specified by the URDF. I then modified `compareParsers.m` to catch this exception and skip the URDF file with a world link.

I also added unit test `drake/systems/plants/test/world_link_test.m`to cover the newly added exception. 

The two unit tests can be run by executing:

```
$ ctest -VV -R systems/plants/test/world_link_test
$ ctest -VV -R systems/plants/test/compareParsers
```

---

*[drake/systems/plants/@RigidBodyManipulator/RigidBodyManipulator.m, line 355 \[r12\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KIF2qS3p2IElolLD5nm:-KIFv7cO17nMma6QyyCO:-160586750) ([raw file](https://github.com/robotlocomotion/drake/blob/b645fd1d5ae0704779dda2640a0bab308ac7aa16/drake/systems/plants/@RigidBodyManipulator/RigidBodyManipulator.m#L355)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> Note that this can be written to be more efficient without the for loop.  But it's fine for this PR.
</details>
I ended up removing this method due to a major bug (see my other comment).

I am curious how it can be done more efficiently given that the link names are not sorted.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-:-KIGaifTae-evML5Zcws:1687017439)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Reviewed 7 of 7 files at r13.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[drake/systems/plants/RigidBodyTreeSDF.cpp, line 218 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KIDYnkszFwIjNsVeI_Q:-KIIA3lrIBfNbICjfWB1:-1810545291) ([raw file](https://github.com/robotlocomotion/drake/blob/ceefe8a3a2422a4e665169fc22e2efca46e2aa1d/drake/systems/plants/RigidBodyTreeSDF.cpp#L218)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Fixed. I forgot about the availability of those macros. Thanks!
</details>
BTW Using __FILE__ places the full build path into the error message, e.g., the message will be "/home/jwnimmer/jwnimmer-tri/drake-distro/drake/../drake/systems/plants/RigidBodyTreeSDF.cpp: parseSDFLink: ERROR: Link tag...".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/systems/plants/RigidBodyTreeSDF.cpp, line 218 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KIDYnkszFwIjNsVeI_Q:-KIICcSpbqptzRs6ssVi:-868054011) ([raw file](https://github.com/robotlocomotion/drake/blob/ceefe8a3a2422a4e665169fc22e2efca46e2aa1d/drake/systems/plants/RigidBodyTreeSDF.cpp#L218)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Using __FILE__ places the full build path into the error message, e.g., the message will be "/home/jwnimmer/jwnimmer-tri/drake-distro/drake/../drake/systems/plants/RigidBodyTreeSDF.cpp: parseSDFLink: ERROR: Link tag...".
</details>
Interesting. While that may not be a defect, since the info _is_ useful, do you know what is the best practice in terms of extracting just the file name?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[drake/systems/plants/RigidBodyTreeSDF.cpp, line 218 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KIDYnkszFwIjNsVeI_Q:-KIIDSEE7dcRKZ4f6qXN:2124924175) ([raw file](https://github.com/robotlocomotion/drake/blob/ceefe8a3a2422a4e665169fc22e2efca46e2aa1d/drake/systems/plants/RigidBodyTreeSDF.cpp#L218)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Interesting. While that may not be a defect, since the info _is_ useful, do you know what is the best practice in terms of extracting just the file name?
</details>
FYI Going from fullname to basename would typically be a call to a filename-handling helper function (so using `tinydir`, in our case).  Sometimes, projects also choose to retain the source-relative path, so in this case show `drake/systems/plants/RigidBodyTreeSDF.cpp: parseSDFLink: ...`, i.e., just strip the build location (the jwnimmer stuff).  Many projects will also try to factor out a lot of this boilerplate, so that error-generating callsites mostly only focus on the error message; so e.g. `if (!attr) { DRAKE_THROW() << "Do not name a ..."; }`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[drake/systems/plants/RigidBodyTreeSDF.cpp, line 218 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KIDYnkszFwIjNsVeI_Q:-KIIJKkIlL_koI9VRgDA:128106795) ([raw file](https://github.com/robotlocomotion/drake/blob/ceefe8a3a2422a4e665169fc22e2efca46e2aa1d/drake/systems/plants/RigidBodyTreeSDF.cpp#L218)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> FYI Going from fullname to basename would typically be a call to a filename-handling helper function (so using `tinydir`, in our case).  Sometimes, projects also choose to retain the source-relative path, so in this case show `drake/systems/plants/RigidBodyTreeSDF.cpp: parseSDFLink: ...`, i.e., just strip the build location (the jwnimmer stuff).  Many projects will also try to factor out a lot of this boilerplate, so that error-generating callsites mostly only focus on the error message; so e.g. `if (!attr) { DRAKE_THROW() << "Do not name a ..."; }`.
</details>
BTW I forgot that `tinydir` doesn't actually have path-manipulation routines.  So I guess for now we'd need to implement `basename` by hand if we want it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/systems/plants/RigidBodyTreeSDF.cpp, line 218 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KIDYnkszFwIjNsVeI_Q:-KIIGe0yHS1vxQ2xjOWL:-219434731) ([raw file](https://github.com/robotlocomotion/drake/blob/ceefe8a3a2422a4e665169fc22e2efca46e2aa1d/drake/systems/plants/RigidBodyTreeSDF.cpp#L218)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW I forgot that `tinydir` doesn't actually have path-manipulation routines.  So I guess for now we'd need to implement `basename` by hand if we want it.
</details>
@sherm1, would you be opposed to the introduction of a `DRAKE_THROW()` macro? Maybe in `drake/common`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171)*
<!-- Sent from Reviewable.io -->

liangfok commented:
The CI build errors appear to be legit. Investigating...

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> @ggould-tri and @russtedrake: I've addressed your review comments. Thanks for the reviews. PTAL.
</details>


---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-:-KIIT7xKjYPdxclu0Rb-:-649596905)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I'm not sure why, but I suspect my refactor from using `compare()` to `==` between strings resulted in the unit tests failing. I've pushed the fix. Hopefully CI will pass now.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> The CI build errors appear to be legit. Investigating...
</details>


---

Review status: 15 of 16 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-:-KIIm20m29bbMYT9Szh4:298284426)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/systems/plants/RigidBodyTreeSDF.cpp, line 218 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KIDYnkszFwIjNsVeI_Q:-KIIw1pX3fcLA0UvZrkv:541929128) ([raw file](https://github.com/robotlocomotion/drake/blob/ceefe8a3a2422a4e665169fc22e2efca46e2aa1d/drake/systems/plants/RigidBodyTreeSDF.cpp#L218)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> @sherm1, would you be opposed to the introduction of a `DRAKE_THROW()` macro? Maybe in `drake/common`?
</details>
FYI I suspect that a separate PR, with a separate design discussion, would be the most effective way to make progress on the "reusable throws" infrastructure question.  IMO, it is intertwined with how to phrase assertions (#1935) -- I think that DRAKE_THROW, DRAKE_ASSERT, DRAKE_ASSERT_ALWAYS, and DRAKE_CHECK (my proposal in the works) overlap in the design space, and a single design proposal to add them all to `common` at once would make the debate most effective.  I can volunteer to write up a sample PR very soon, if that helps.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
*[drake/systems/plants/RigidBodyTreeSDF.cpp, line 218 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KIDYnkszFwIjNsVeI_Q:-KIKOnT8oXWV376cLrO5:-578014469) ([raw file](https://github.com/robotlocomotion/drake/blob/ceefe8a3a2422a4e665169fc22e2efca46e2aa1d/drake/systems/plants/RigidBodyTreeSDF.cpp#L218)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> FYI I suspect that a separate PR, with a separate design discussion, would be the most effective way to make progress on the "reusable throws" infrastructure question.  IMO, it is intertwined with how to phrase assertions (#1935) -- I think that DRAKE_THROW, DRAKE_ASSERT, DRAKE_ASSERT_ALWAYS, and DRAKE_CHECK (my proposal in the works) overlap in the design space, and a single design proposal to add them all to `common` at once would make the debate most effective.  I can volunteer to write up a sample PR very soon, if that helps.
</details>
I agree, @jwnimmer-tri. Also might be worth thinking about how this relates to logging in #1895. Spdlog has TRACE and DEBUG macros that have a particular formatting style; likely not something we want to replicate but worth a look. Useful features of having our own macros: inclusion of FILE, LINE, func information; ability to control behavior with compile time flags; help formatting error messages. I've seen these with printf-style formatting and with ostringstream-style.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171)*
<!-- Sent from Reviewable.io -->

liangfok commented:
[CI experimental build #1617](https://drake-jenkins.csail.mit.edu/job/experimental/1617/) appears to have suffered the same fate with a stalled `clang-mac` test.

The Mac CI build servers appear to be offline: https://drake-jenkins.csail.mit.edu/computer/

![Screen Shot 2016-05-21 at 6.28.24 PM.png](https://files.reviewableusercontent.io/images/reviewable/1388098/gtlxGJKD7wQQ3QXDNtUkN78i/Screen%20Shot%202016-05-21%20at%206.28.24%20PM.png)

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> [CI experimental build #1615](https://drake-jenkins.csail.mit.edu/job/experimental/1615/) appears to have stalled on `clang-mac` with the following error:
> 
> ```
> 15:29:01 Configuration experimental » clang,mac is still in the queue: All nodes of label ‘mac’ are offline
> ```
> 
> I'm going to try triggering another round of CI experimental build now.
> 
> @drake-jenkins-bot retest this please
</details>


---

Review status: 15 of 16 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-:-KIKGxwogxb13v1G8ddc:1134076700)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
This is <img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> for C++ platform review.  I am not qualified to comment on the platform qualities of the matlab; it seems like Russ is on that.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> [CI experimental build #1617](https://drake-jenkins.csail.mit.edu/job/experimental/1617/) appears to have suffered the same fate with a stalled `clang-mac` test.
> 
> The Mac CI build servers appear to be offline: https://drake-jenkins.csail.mit.edu/computer/
> 
> ![Screen Shot 2016-05-21 at 6.28.24 PM.png](https://files.reviewableusercontent.io/images/reviewable/1388098/gtlxGJKD7wQQ3QXDNtUkN78i/Screen%20Shot%202016-05-21%20at%206.28.24%20PM.png)
> 
</details>


---

Reviewed 1 of 7 files at r13, 1 of 1 files at r14.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 589 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KIDb97qVPGfskaDO7UT:-KISf2VQ1WTzZhNoZctd:-2012483613) ([raw file](https://github.com/robotlocomotion/drake/blob/ceefe8a3a2422a4e665169fc22e2efca46e2aa1d/drake/systems/plants/RigidBodyTreeURDF.cpp#L589)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I've updated the documentation.
> 
> It is an implicit specification of URDF. Each URDF file describes one robot [1] using a tree of links connected by joints. Thus, the only way for a URDF to contain two world-connecting joints is if the URDF describes more than a one robot. This is a violation of the one-robot-per-URDF rule.
> 
> [1] http://answers.gazebosim.org/question/62/sdf-vs-urdf-what-should-one-use/
</details>
I'm marking this satisfied because it is clear that you understand better than I do.  However I am still confused by your explanation.  Even in a tree a single node can have many edges; you seem to be saying that world must be a leaf in the tree.  Is it the case that in any given robot tree the world is always a leaf?

Perhaps we should chat face-to-face some time to clear up my confusion.  In the meantime, there is no reason this should be a barrier to merging.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-:-KISg8rf2YUVUiGmjpDk:1454502660)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 589 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-KIDb97qVPGfskaDO7UT:-KIShKcepgcoqNudxQTI:-1105793378) ([raw file](https://github.com/robotlocomotion/drake/blob/ceefe8a3a2422a4e665169fc22e2efca46e2aa1d/drake/systems/plants/RigidBodyTreeURDF.cpp#L589)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> I'm marking this satisfied because it is clear that you understand better than I do.  However I am still confused by your explanation.  Even in a tree a single node can have many edges; you seem to be saying that world must be a leaf in the tree.  Is it the case that in any given robot tree the world is always a leaf?
> 
> Perhaps we should chat face-to-face some time to clear up my confusion.  In the meantime, there is no reason this should be a barrier to merging.
</details>
I think I understand the confusion. I'm referring to the world link _within_ a single URDF file whereas you're referring to the world link within Drake's `RigidBodyTree`. Is this correct?

The "world" link is always the root of the tree. Within Drake's `RigidBodyTree`, there can be multiple robots attached to the world. However, within a single URDF file, since there is only one robot modeled per URDF file, the world link is always attached via a single (typically 6DOF floating) joint to the root of the robot model.

Does the above make sense?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
@drake-jenkins-bot retest this please

jamiesnape commented:
@drake-jenkins-bot retest this please

liangfok commented:
This PR passed [CI Build 1632](https://drake-jenkins.csail.mit.edu/job/experimental/1632/). Thus, +@russtedrake for review of Matlab code.

<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> This is :LGTM: for C++ platform review.  I am not qualified to comment on the platform qualities of the matlab; it seems like Russ is on that.
</details>


---

Reviewed 1 of 7 files at r13.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171#-:-KITk_L9wMNF1zjAX7Gm:1586909230)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Reviewed 1 of 1 files at r14.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2171)*
<!-- Sent from Reviewable.io -->
